Makefile: misc improvements

knz opened the pull request:
1. both `find` invocations to clean up stale generated files can
   redirect errors to stderr.

2. the `.tmp` file produced by `execgen >.tmp` must be removed if
   `execgen` encounters an error.

3. the `execgen` + `cmp` + `mv/rm` invocation was very wide and
   verbose. Let's make it shorter.

4. there was no need to duplicate the dep graph between `execgen` and `Makefile`. Let's simplify.

5. it's a little bit sad (although functionally 100% correct) that the `execgen` targets must run again on every `make` invocation. To make the build a tad more compact, this can be simplified too.


Release note: None

knz committed:
Makefile: misc improvements

1. both `find` invocations to clean up stale generated files can
   redirect errors to stderr.

2. the `.tmp` file produced by `execgen >.tmp` must be removed if
   `execgen` encounters an error.

3. the `execgen` + `cmp` + `mv/rm` invocation was very wide and
   verbose. Let's make it shorter.

Note: this current structure still mandates to run `execgen` in any
case, because if the `_tmpl.go` file is newer than the `.eg.go`
but the `.eg.go` doesn't get an update, `make` will always keep
thinking the generated file needs updating. This will be looked at in
the next patch.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/45797)
<!-- Reviewable:end -->

knz committed:
Makefile,execgen: de-dup the dependency list

Prior to this patch, the dependency graph for `execgen` was contained
both in the `execgen` executable itself and the `Makefile`.
This was error-prone.

This patch changes it to have `execgen` as primary source of truth
and export its dependency graph for use by `Makefile.`

Release note: None

knz committed:
Makefile: avoid running `execgen` on every `make` invocation

The explanation for the commit is included in a comment in the execgen
rule. Copy below.

This makes it so that two subsequent `make` invocations do not
cause all the `execgen` target to re-run.

```
    # Note how the dependency work is complete after the cmp/rm/mv
    # dance to write to the output.
    # However, because it does not always change the timestamp of
    # the target file, it is possible to get a situation where the
    # target is older than both bin/execgen and also the _tmpl.go
    # file, but does not get changed by this rule. This happens
    # e.g. after a 'git checkout' to a different branch, when the
    # execgen binary and templates do not change across branches.
    #
    # In order to prevent the rule from kicking again in every
    # make invocation, we tweak the timestamp of the output file
    # to be the latest of either bin/execgen or the input template.
    # This makes it just new enough that 'make' will be satisfied
    # that it does not need an update any more.
    #
    # Note that we don't want to ratchet the timestamp all the way
    # to the present, because then it will becomes newer
    # than all the other produced artifacts downstream and force
    # them to rebuild too.
```

Release note: none

jordanlewis commented:
> it's a little bit sad (although functionally 100% correct) that the execgen targets must run again on every make invocation. To make the build a tad more compact, this can be simplified too.

What is the simplification you are implying here?

I can't think of a way to avoid running execgen on every make invocation except to revert back to what we had, which I think realistically might be better in more cases. Compiling `colexec` itself is not a high cost, and if any of the dependencies of `execgen` change, it's more than likely that `sql` will have to get recompiled again *anyway*.

knz commented:
> What is the simplification you are implying here?

Look at the last commit. It works and does not get us in a worse place than the latest change you made.

jordanlewis commented:
Yeah, by "revert" I meant go back to before my change, not yours, which seems like much better than what I did :)

irfansharif reviewed (commented):
LGTM except for the last commit, which is a bit beyond me. When I looked into why we kept rebuilding execgen, I found a different underlying cause. See comment below.

> Reviewed 1 of 1 files at r1, 34 of 34 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-:-M1kMqGYCd6TVbsSEa2o:b8a8jna)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @jordanlewis, @knz, and @yuzefovich)

___
*[Makefile, line 842 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kD5xHAXS1_cdZS7bN:-M1kD5xHAXS1_cdZS7bO:bw5vukm) ([raw file](https://github.com/cockroachdb/cockroach/blob/9848ab7549e88fa895404d3383aff4dd65cfaf51/Makefile#L842)):*
> ```Makefile
> execgen-exclusions = $(addprefix -not -path ,$(EXECGEN_TARGETS))
> 
> $(info Cleaning old generated files.)
> ```

Would be nice to remove this `info`, I was a bit confused by it seeing it even when there weren't any old generated files to be cleaned up.
___
*[Makefile, line 1539 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kDcZj9MDnPYddUmjt:-M1kDcZj9MDnPYddUmju:bo6vnsb) ([raw file](https://github.com/cockroachdb/cockroach/blob/9848ab7549e88fa895404d3383aff4dd65cfaf51/Makefile#L1539)):*
> ```Makefile
> 
> %.eg.go: bin/execgen
> 	@echo EXECGEN $@; execgen $@ > $@.tmp || { rm -f $@.tmp; exit 1; }
> ```

Break it into next line + `@execgen $@...`?
___
*[Makefile, line 709 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kINH17MbtgQUs3J3Z:-M1kINH17MbtgQUs3J3_:bdr396j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bef04bd0acf6179727b93b2890dd3fa25ff1743e/Makefile#L709)):*
> ```Makefile
> 
> $(LIBPROTOBUF): $(PROTOBUF_DIR)/Makefile bin/uptodate .ALWAYS_REBUILD
> 	@uptodate $@ $(PROTOBUF_SRC_DIR) || $(MAKE) --no-print-directory -C $(PROTOBUF_DIR) libprotobuf
> ```

We need something like `&& touch -m $@` here, I think. I was looking into why `bin/execgen` kept getting rebuilt on my laptop yesterday (through `make buildshort; make buildshort --debug`), and I found that it was because of the `LIBPROTOBUF` target.  My hypothesis was: `uptodate` checks if all mtimes of files in the src dir are less than the mtime of the target itself, but on my computer this was not the case for some random file that didn't affect the generated `libprotobuf.a`. But because we defer to the `Makefile` in the protobuf src directory, which correctly assesses that `libprotobuf.a` doesn't need to get rebuilt, it's `mtime` is left unchanged. This therefore causes all subsequent `uptodate` calls to fail, as we never change anything.

But hm, after having typed the above out and seeing your work below, I realize I might have been wrong in that it's _this_ issue that's causing `bin/execgen` to keep getting rebuilt.
___
*[Makefile, line 1509 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kGQrs0X4r7TbehD2E:-M1kGQrs0X4r7TbehD2F:b-vha67s) ([raw file](https://github.com/cockroachdb/cockroach/blob/bef04bd0acf6179727b93b2890dd3fa25ff1743e/Makefile#L1509)):*
> ```Makefile
> # However, because it does not always change the timestamp of
> # the target file, it is possible to get a situation where the
> # target is older than both bin/execgen and also the _tmpl.go
> ```

s/also//
___
*[pkg/sql/colexec/execgen/cmd/execgen/main.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kFNkaEeb_BUwKInNo:-M1kFNkaEeb_BUwKInNp:bbips1e) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dbfeab436fef3eefce00a865f4aa234d8de0c6d/pkg/sql/colexec/execgen/cmd/execgen/main.go#L89)):*
> ```Go
> 	for _, out := range args {
> 		_, file := filepath.Split(out)
> 		gen := generators[file]
> ```

s/gen/entry


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
>  When I looked into why we kept rebuilding execgen, I found a different underlying cause. See comment below.

These are two different problems. Even if `execgen` does not need to get rebuilt, the .eg.go get re-processed currently. The reason is explained in my comment.

Your proposal re protobuf ensures that `execgen` does not get rebuilt, but it's an orthogonal concern.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-:-M1kQ0dF1_VRBKvFn_0B:b-lqztmz)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @irfansharif, @jordanlewis, @knz, and @yuzefovich)



<!-- Sent from Reviewable.io -->

petermattis reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-:-M1kQi2mDs56xdlGwjkc:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif, @jordanlewis, @knz, and @yuzefovich)

___
*[Makefile, line 1501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kPM5Q3COHXUteHYhD:-M1kPM5Q3COHXUteHYhE:b-8bsg4j) ([raw file](https://github.com/cockroachdb/cockroach/blob/bef04bd0acf6179727b93b2890dd3fa25ff1743e/Makefile#L1501)):*
> ```Makefile
> 	@mv -f $@.tmp $@
> 
> include bin/execgen_out.d
> ```

I believe this causes the `bin/execgen_out.d` rule to be run. I always forget what make does when it sees the pattern rules below. I hope it isn't regenerating them as well, but I'm not certain just looking at this.

I'm not clear what `bin/execgen_out.d` is doing and I don't see mention of it in the very nice comment below.

Update: Ah, just discovered the new `-M` option. I think it is worth adding a comment above the `bin/execgen_out.d` rule. And you can ignore my comments above.
___
*[Makefile, line 1533 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kQ17d23kjLI94QaI_:-M1kQ17d23kjLI94QaIa:bpkcbiy) ([raw file](https://github.com/cockroachdb/cockroach/blob/bef04bd0acf6179727b93b2890dd3fa25ff1743e/Makefile#L1533)):*
> ```Makefile
> 	    target_timestamp_file=bin/execgen; \
> 	  fi; \
> 	  touch -r $$target_timestamp_file $@
> ```

This is complicated, but this is the way of make. You can do anything with make, but it is like writing in assembly language.


<!-- Sent from Reviewable.io -->

knz committed:
Makefile: remove the "Cleaning up" message

If there is no file to clean up, the "Cleaning up" message
should not be printed.

Release note: None

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45797)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @irfansharif, @jordanlewis, @knz, @petermattis, and @yuzefovich)

___
*[Makefile, line 842 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kD5xHAXS1_cdZS7bN:-M1kcbJNAb3sz42p0LOW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9848ab7549e88fa895404d3383aff4dd65cfaf51/Makefile#L842)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Would be nice to remove this `info`, I was a bit confused by it seeing it even when there weren't any old generated files to be cleaned up.
</blockquote></details>

Done.
___
*[Makefile, line 1539 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kDcZj9MDnPYddUmjt:-M1kctn46tURKDB-X5Pe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/9848ab7549e88fa895404d3383aff4dd65cfaf51/Makefile#L1539)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Break it into next line + `@execgen $@...`?
</blockquote></details>

Done.
___
*[Makefile, line 1501 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kPM5Q3COHXUteHYhD:-M1kcchEAEiw6B0QcxRF:b-u5t80l) ([raw file](https://github.com/cockroachdb/cockroach/blob/bef04bd0acf6179727b93b2890dd3fa25ff1743e/Makefile#L1501)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I believe this causes the `bin/execgen_out.d` rule to be run. I always forget what make does when it sees the pattern rules below. I hope it isn't regenerating them as well, but I'm not certain just looking at this.

I'm not clear what `bin/execgen_out.d` is doing and I don't see mention of it in the very nice comment below.

Update: Ah, just discovered the new `-M` option. I think it is worth adding a comment above the `bin/execgen_out.d` rule. And you can ignore my comments above.
</blockquote></details>

Added comment to clarify.
___
*[Makefile, line 1509 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kGQrs0X4r7TbehD2E:-M1kcqpt86oCLUOVt1aj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/bef04bd0acf6179727b93b2890dd3fa25ff1743e/Makefile#L1509)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/also//
</blockquote></details>

Done.
___
*[Makefile, line 1533 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kQ17d23kjLI94QaI_:-M1kdCuz6FW7ZXC9le2H:b-ngm0ln) ([raw file](https://github.com/cockroachdb/cockroach/blob/bef04bd0acf6179727b93b2890dd3fa25ff1743e/Makefile#L1533)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

This is complicated, but this is the way of make. You can do anything with make, but it is like writing in assembly language.
</blockquote></details>

I would dream of a world where the assembly language is the unix shell. Seriously, it's not so bad.
___
*[pkg/sql/colexec/execgen/cmd/execgen/main.go, line 89 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kFNkaEeb_BUwKInNo:-M1kdm00CN-cJ5pKyEAA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8dbfeab436fef3eefce00a865f4aa234d8de0c6d/pkg/sql/colexec/execgen/cmd/execgen/main.go#L89)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

s/gen/entry
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

knz reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/45797)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @irfansharif, @jordanlewis, @petermattis, and @yuzefovich)

___
*[Makefile, line 709 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/45797#-M1kINH17MbtgQUs3J3Z:-M1kdv_N1kom32o64NHt:b-96j3wa) ([raw file](https://github.com/cockroachdb/cockroach/blob/bef04bd0acf6179727b93b2890dd3fa25ff1743e/Makefile#L709)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

We need something like `&& touch -m $@` here, I think. I was looking into why `bin/execgen` kept getting rebuilt on my laptop yesterday (through `make buildshort; make buildshort --debug`), and I found that it was because of the `LIBPROTOBUF` target.  My hypothesis was: `uptodate` checks if all mtimes of files in the src dir are less than the mtime of the target itself, but on my computer this was not the case for some random file that didn't affect the generated `libprotobuf.a`. But because we defer to the `Makefile` in the protobuf src directory, which correctly assesses that `libprotobuf.a` doesn't need to get rebuilt, it's `mtime` is left unchanged. This therefore causes all subsequent `uptodate` calls to fail, as we never change anything.

But hm, after having typed the above out and seeing your work below, I realize I might have been wrong in that it's _this_ issue that's causing `bin/execgen` to keep getting rebuilt.
</blockquote></details>

@irfansharif I'm not going to change this in this PR. Can you make another PR based off this one and try it out? 


<!-- Sent from Reviewable.io -->

irfansharif commented:
Will do. Thanks for sending this in.

irfansharif commented:
I'm not sure how, but I think due to this change `make clean` does not do what you'd expect it to.

```sh
$ make clean
[...]
[ 36%] Building CXX object CMakeFiles/libprotoc.dir/Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/c-deps/protobuf/src/google/protobuf/compiler/java/java_service.cc.o
[ 37%] Building CXX object CMakeFiles/libprotoc.dir/Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/c-deps/protobuf/src/google/protobuf/compiler/java/java_shared_code_generator.cc.o
[ 37%] Building CXX object CMakeFiles/libprotoc.dir/Users/irfansharif/Software/src/github.com/cockroachdb/cockroach/c-deps/protobuf/src/google/protobuf/compiler/java/java_string_field.cc.o
```
